packages/flutter_tools/lib/src/build_system/source.dart

```diff
@@ -26,16 +38,55 @@ class SourceVisitor {
   /// Defaults to `true`.
   final bool inputs;
 
-  /// The entities are populated after visiting each source.
+  @override
   final List<File> sources = <File>[];
 
+  @override
+  bool get containsNewDepfile => _containsNewDepfile;
+  bool _containsNewDepfile = false;
+
   /// Visit a [Source] which contains a function.
   ///
   /// The function is expected to produce a list of [FileSystemEntities]s.
   void visitFunction(InputFunction function) {
     sources.addAll(function(environment));
   }
 
+  /// Visit a depfile which contains both input and output files.
+  ///
+  /// If the file is missing, this visitor is marked as [containsNewDepfile].
+  /// This is used by the [Node] class to tell the [BuildSystem] to
+  /// defer hash computation until after executing the target.
+  void visitDepfile(String name) {
+    final File depfile = environment.buildDir.childFile(name);
+    if (!depfile.existsSync()) {
+      _containsNewDepfile = true;
+      return;
+    }
+    final String contents = depfile.readAsStringSync();
+    final List<String> colonSeparated = contents.split(': ');
+    if (inputs) {
+      sources.addAll(_processList(colonSeparated[1].trim()));
+    } else {
+      sources.addAll(_processList(colonSeparated[0].trim()));
+    }
+  }
+
+  final RegExp _separatorExpr = RegExp(r'([^\\]) ');
+  final RegExp _escapeExpr = RegExp(r'\\(.)');
+
+  Iterable<File> _processList(String rawText) {
```

Comment:
This processing needs more comprehensive testing, esp. around the escape sequences. Some examples here in the comments of this function would also be good.

Reply:
I've added a windows path example and a spaces in path example. Also linked to the code that I adapted this from